### PR TITLE
Reuse dynamic filters

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java [hunk 1]
```diff
@@ -578,10 +578,19 @@ private DynamicFiltersResult createDynamicFilters(JoinNode node, List<JoinNode.E
...
                     Symbol buildSymbol = clause.getRight();
-                    String id = "df_" + idAllocator.getNextId().toString();
+                    String id;
+                    if (buildSymbolToDynamicFilter.containsKey(buildSymbol)) {
+                        id = buildSymbolToDynamicFilter.get(buildSymbol);
+                    }
+                    else {
+                        id = "df_" + idAllocator.getNextId().toString();
+                        buildSymbolToDynamicFilter.put(buildSymbol, id);
+                    }
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java:593]:
```suggestion
                    id = buildSymbolToDynamicFilter.computeIfAbsent(buildSymbol, key -> "df_" + idAllocator.getNextId().toString());
```

However, it's not a good idea to modify the map returned by `.collect(toMap(...))`. There's no guarantee that the map is mutable or that it's safe to change. From the javadocs:

> There are no guarantees on the type, mutability, serializability, or thread-safety of the Map returned.

**Comment by findepi** [comment]
> I do not see a benefit for that. What is the motivation?

Let's capture this in a code comment.

**Comment by sopel39** [comment]
> LGTM generally though I do not see a benefit for that. What is the motivation?

We generally want plans to be stable (don't use new symbols or node ids if possible). Additionally, it could be the case (hypothetically) that same build symbol is used with different probe symbols. In such case, you want to have just one dynamic filter.

**Comment by sopel39** [comment]
ac @martint

**Pull request merged**